src/main/kotlin/org/jqassistant/tooling/intellij/plugin/editor/rules/UastItReferenceContributor.kt

```diff
@@ -0,0 +1,131 @@
+import org.jetbrains.uast.getUCallExpression
+import org.jetbrains.uast.resolveToUElementOfType
+import org.jetbrains.uast.wrapULiteral
+import org.jqassistant.tooling.intellij.plugin.data.rules.JqaRuleType
+
+/**
+ * Injects soft references to jQA rules for integration tests written in
+ * UAST languages (e.g. Kotlin, Java).
+ */
+class UastItReferenceContributor : PsiReferenceContributor() {
+    override fun registerReferenceProviders(registrar: PsiReferenceRegistrar) {
+        registrar.registerUastReferenceProvider(
+            injectionHostUExpression().filterWithContext { argumentExpression, processingContext ->
+                val call =
+                    argumentExpression.uastParent.getUCallExpression(searchLimit = 2) ?: return@filterWithContext false
+
+                val constructorOrMethodCall = setOf(UastCallKind.CONSTRUCTOR_CALL, UastCallKind.METHOD_CALL)
+
+                // All possible literals of the function call
+                val argumentLiterals = call.valueArguments.map(::wrapULiteral)
+                // The literal we are currently checking
+                val searchingLiteral = wrapULiteral(argumentExpression)
+
+                // Get the index of the current parameter we are checking
+                val annotatedParameterIndex = argumentLiterals.indexOfFirst { lit -> lit == searchingLiteral }
+                // This should never happen
+                if (annotatedParameterIndex == -1) return@filterWithContext false
+
+                // Tell the Reference provider which parameters can be autocompleted
+                processingContext.put("annotatedParameterLiterals", argumentLiterals)
+                // Tell the AnnotationPatternCondition which of
+                // the parameters needs to have the annotation
+                processingContext.put("annotatedParameterIndex", annotatedParameterIndex)
+
+                // This checks if the current parameter has the annotation
+                val callPattern = callExpression().with(AnnotationPatternCondition)
+                callPattern.accepts(call, processingContext) &&
+                    call.kind in constructorOrMethodCall
+            },
+            UastItReferenceProvider,
+        )
+    }
+}
+
+/**
+ * Checks if a method call contains a correctly annotated parameter at the `annotatedParameterIndex`
+ * that is set in the current ProcessingContext
+ */
+object AnnotationPatternCondition : PatternCondition<UCallExpression>("jQARuleIdentifierAnnotationPattern") {
+    override fun accepts(t: UCallExpression, context: ProcessingContext?): Boolean {
+        /*
+        // This is twice as fast for some reason
+        val psiElement = t.sourcePsi ?: return false
+        val firstChild = psiElement.firstChild
+        val methodReference = firstChild.reference ?: return false
+        val methodElement = methodReference.resolve()
+        val uMethodElement = methodElement.toUElement(UMethod::class.java) ?: return false
```

Comment:
The slowdown is 0.8ms instead of 0.4ms, but given the strongly increased readability I think it is worth the minor (for me unnoticeable) performance cost